providers/implementations/signature/rsa_sig.c

```diff
@@ -67,32 +75,73 @@ static OSSL_ITEM padding_item[] = {
     { 0,                        NULL     }
 };
 
+/*
+ * Internal hash function method.  For the non-sigalg implementation,
+ * this is just a wrapper around EVP_MD functionality.  For sigalg
+ * implementations, however, a more direct approach is taken, which
+ * depends entirely on what hash algorithms are implemented in this
+ * provider.
+ */
+struct md_wrapper_st {
+    int mdnid;                       /* Note: hard-coded by sigalg impl */
+    char mdname[OSSL_MAX_NAME_SIZE]; /* Purely informational */
+    size_t mdsize;                   /* Cached */
+
+    /* |md| is reserved for non-sigalg impls */
+    EVP_MD *md;
+
+    /* The OSSL_FUNC_digest pointers are reserved for sigalg impls */
+    OSSL_FUNC_digest_newctx_fn                  *newctx;
+    OSSL_FUNC_digest_freectx_fn                 *freectx;
+    OSSL_FUNC_digest_dupctx_fn                  *dupctx;
+    OSSL_FUNC_digest_init_fn                    *init;
+    OSSL_FUNC_digest_digest_fn                  *digest;
+    OSSL_FUNC_digest_update_fn                  *update;
+    OSSL_FUNC_digest_final_fn                   *final;
+    OSSL_FUNC_digest_get_params_fn              *get_params;
+    OSSL_FUNC_digest_settable_ctx_params_fn     *settable_ctx_params;
+    OSSL_FUNC_digest_set_ctx_params_fn          *set_ctx_params;
+    OSSL_FUNC_digest_gettable_ctx_params_fn     *gettable_ctx_params;
+    OSSL_FUNC_digest_get_ctx_params_fn          *get_ctx_params;
+};
+
 /*
  * What's passed as an actual key is defined by the KEYMGMT interface.
  * We happen to know that our KEYMGMT simply passes RSA structures, so
  * we use that here too.
  */
 
 typedef struct {
+    void *provctx;
     OSSL_LIB_CTX *libctx;
     char *propq;
     RSA *rsa;
     int operation;
 
+    /*
+     * Flag to determine if a full sigalg is run (1) or if a composable
+     * signature algorithm is run (0).
+     *
+     * When a full sigalg is run (1), this currently affects the following
+     * other flags, which are to remain untouched after their initialization:
+     *
+     * - flag_allow_md (initialized to 0)
+     */
+    unsigned int flag_sigalg : 1;
```

Comment:
Would `flag_full_sigalg` be better?

Reply:
In the numenclature we already use, sigalg *is* a full signature algorithm...

Reply:
confusing... but I do not insist.